xonsh/lexer.py

```diff
@@ -1,13 +1,307 @@
+    state['pymode'].pop()
+    state['last'] = token
+    yield _new_token('RPAREN', ')', token.start)
+
+
+def handle_rbrace(state, token, stream):
+    state['pymode'].pop()
+    state['last'] = token
+    yield _new_token('RBRACE', '}', token.start)
+
+
+def handle_rbracket(state, token, stream):
+    state['pymode'].pop()
+    state['last'] = token
+    yield _new_token('RBRACKET', ']', token.start)
+
+
+def handle_error_space(state, token, stream):
+    if not state['pymode'][-1]:
+        state['last'] = token
+        yield _new_token('WS', ' ', token.start)
+    else:
+        yield from []
+
+special_handlers = {
+    tokenize.ENCODING: lambda s, t, st: [],
+    tokenize.COMMENT: lambda s, t, st: [],
+    tokenize.ENDMARKER: lambda s, t, st: [],
+    tokenize.NEWLINE: handle_newline,
+    (tokenize.OP, '('): handle_lparen,
+    (tokenize.OP, ')'): handle_rparen,
+    (tokenize.OP, '['): handle_lbracket,
+    (tokenize.OP, ']'): handle_rbracket,
+    (tokenize.OP, '{'): handle_lbrace,
+    (tokenize.OP, '}'): handle_rbrace,
+    (tokenize.ERRORTOKEN, '$'): handle_dollar,
+    (tokenize.ERRORTOKEN, '`'): handle_backtick,
+    (tokenize.ERRORTOKEN, '?'): handle_question,
+    (tokenize.OP, '@'): handle_at,
+    (tokenize.ERRORTOKEN, ' '): handle_error_space,
+    tokenize.INDENT: handle_indent,
+    tokenize.DEDENT: handle_indent,
+}
+
+
+def handle_token(state, token, stream):
+    typ = token.type
+    st = token.string
+    if not state['pymode'][-1]:
+        if state['last'] is not None and state['last'].end != token.start:
+            cur = token.start
+            old = state['last'].end
+            yield _new_token('WS', ' '*(cur[1]-old[1]), old)
+    if (typ, st) in token_map:
+        state['last'] = token
+        yield _new_token(token_map[(typ, st)], st, token.start)
+    elif typ in token_map:
+        state['last'] = token
+        yield _new_token(token_map[typ], st, token.start)
+    elif (typ, st) in special_handlers:
+        for i in special_handlers[(typ, st)](state, token, stream):
+            yield i
+    elif typ in special_handlers:
+        for i in special_handlers[typ](state, token, stream):
+            yield i
+    else:
+        m = "Unexpected token: {0}".format(token)
+        yield _new_token("ERRORTOKEN", m, token.start)
+
+
+def preprocess_tokens(tokstream):
+    tokstream = clear_NL(tokstream)
+    state = {'indents': [0], 'pymode': [True], 'last': None}
+    for token in tokstream:
+        for i in handle_token(state, token, tokstream):
+            yield i
+
+
+def clear_NL(tokstream):
+    for i in tokstream:
+        if i.type != tokenize.NL:
+            yield i
+
+
+def single_error(exc):
+    m = "{} (line {}, column {})".format(exc.msg, exc.lineno, exc.offset)
+    yield _new_token("ERRORTOKEN", m, (0, 0))
+
+
+def tok(s):
+    try:
+        return iter(tokenize.tokenize(BytesIO(s.encode('utf-8')).readline))
+    except Exception as e:
+        return iter(single_error(e))
```

Comment:
The problem is that some things (conveniently things that are hard to catch) cause `tokenize.tokenize` to throw an exception (like dedent not matching an indent level, or similar).  In those cases, so that we can report the error easily, I just wanted to yield a single `ERRORTOKEN` so that we would get a consistent syntax error out of the parser.